src/handlers.js

```diff
@@ -22,76 +23,55 @@ var helpers     = require('./helpers');
 var Handlers = function(options) {
   // Validate options
   assert(options.IndexedTask, 'A subclass of data.IndexedTask is required');
-  assert(options.Namespace,   'A subclass of data.Namespace is required');
-  assert(options.queue,       'An instance of taskcluster.Queue is required');
+  assert(options.Namespace, 'A subclass of data.Namespace is required');
+  assert(options.queue, 'An instance of taskcluster.Queue is required');
   assert(options.queueEvents instanceof taskcluster.QueueEvents,
     'An instance of taskcluster.QueueEvents is required');
   assert(options.credentials, 'credentials must be provided');
-  assert(options.routePrefix,       'routePrefix is required');
-  assert(options.monitor,           'monitor is required for statistics');
+  assert(options.routePrefix, 'routePrefix is required');
+  assert(options.monitor, 'monitor is required for statistics');
   // Store options on this for use in event handlers
-  this.IndexedTask      = options.IndexedTask;
-  this.Namespace        = options.Namespace;
-  this.queue            = options.queue;
-  this.queueEvents      = options.queueEvents;
-  this.credentials      = options.credentials;
-  this.routePrefix      = options.routePrefix;
-  this.queueName        = options.queueName;  // Optional
-  this.monitor          = options.monitor;
-  this.listener         = null;
+  this.IndexedTask = options.IndexedTask;
+  this.Namespace = options.Namespace;
+  this.queue = options.queue;
+  this.queueEvents = options.queueEvents;
+  this.credentials = options.credentials;
+  this.routePrefix = options.routePrefix;
+  this.queueName = options.queueName;  // Optional
+  this.monitor = options.monitor;
+  this.pc = null;
+  // Binding for completed tasks
+  this.binding = options.queueEvents.taskCompleted('route.' + options.routePrefix + '.#');
+  this.pulseClient = options.pulseClient;
 };
 
 /** Setup handlers and start listening */
-Handlers.prototype.setup = function() {
-  assert(this.listener === null, 'Cannot setup twice!');
-  var that = this;
+Handlers.prototype.setup = async function() {
+  assert(this.pc === null, 'Cannot setup twice!');
 
   // Create regular expression for parsing routes
   this.routeRegexp = new RegExp('^' + this.routePrefix + '\\.(.*)$');
 
-  // Create listener
-  this.listener = new taskcluster.PulseListener({
-    credentials:          this.credentials,
-    queueName:            this.queueName,
-  });
-
-  // Binding for completed tasks
-  var completedBinding = this.queueEvents.taskCompleted(
-    'route.' + this.routePrefix + '.#'
+  this.pc = await consume(
+    {
+      client: this.pulseClient,
+      bindings: [this.binding],
+      queueName: this.queueName,
+    },
+    message => this.monitor.timedHandler('listener', this.messageHandler(message)),
```

Comment:
I think this has the same issue as described in taskcluster-github.  The call to `this.messageHandler(message)` will be executed and its return value given to `this.monitor.timedHandler`, which will then treat it as taking no time at all to execute.

The way I find most clear to write this is:

```javascript
message => this.monitor.timedHandler('listener', () => this.messageHandler(message)),
```
which is a function taking one argument (message), which calls timedHandler with a handler name ('listener') and a function which takes no arguments and calls `this.messageHandler(message)`.  timedHandler will then call this last function and time how long it takes to execute.

Reply:
Investigating this